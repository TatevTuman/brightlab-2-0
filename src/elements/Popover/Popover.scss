@import "static/assets/styles/mixins";

.popover {
  display: none;
  position: relative;
  cursor: pointer;

  &[data-open='true'] {
    display: block;
  }

  &:after {
    display: block;
    position: absolute;
    content: '';
    transform: rotate(45deg);
    background: #16161b;
    width: toRem(10);
    height: toRem(10);
    left: toRem(120);
    bottom: toRem(-4);
    z-index: -1;
  }

  &-list {
    height: 100%;
    width: toRem(265);
    background: #16161B;
    border-radius: toRem(10);
    box-shadow: 3px 8px 12px rgba(22, 22, 26, 0.251885);
    font-size: toRem(12);
    color: var(--white);
    bottom: toRem(45);
    left: toRem(-120);

    &-line {
      display: inline;
      position: absolute;
      height: 0;
      width: toRem(237);
      bottom: toRem(66.5);
      left: toRem(13);
      border-bottom: toRem(0.5) solid var(--white);
    }

    &-option {
      height: toRem(68);
      padding: toRem(10) toRem(14);
      z-index: 1;

      &:first-child {
        border-radius: toRem(10) toRem(10) 0 0;
      }

      &:last-child {
        border-radius: 0 0 toRem(10) toRem(10);
      }

      &:hover {
        background: rgba(115, 115, 121, 0.5);
      }

      &:active {
        background: rgba(115, 115, 121, 0.82);
      }

      div:first-child {
        mix-blend-mode: normal;
        opacity: 0.7;

        & + div {
          font-weight: 500;

          & + div {
            display: flex;
            height: toRem(12);
            font-weight: 500;

            div:first-child {

              & ~ div {
                padding-left: toRem(7);
              }
            }
          }
        }
      }
    }
  }
}