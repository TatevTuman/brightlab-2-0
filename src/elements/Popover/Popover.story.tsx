import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import Popover, { PopoverProps } from './Popover'
import { action } from '@storybook/addon-actions'
import { storybook } from '@utils'


const options = [
  {
    label: {
      name: 'Michael Stevenson',
      company: 'Figma Corporate. Presentation and Sales...',
      date: '8:00 - 9:30'
    },
    value: 'first Option'
  },
  {
    label: {
      name: 'Michael Stevenson',
      company: 'Figma Corporate. Presentation and Sales...',
      date: '8:00 - 9:30'
    },
    value: 'second Option'
  }
]

export default {
  title: 'Elements/Popover',
  component: Popover,
  argTypes: {
    open: { control: 'boolean', name: 'open' },
    onClick: { action: 'onClick' }
  }
} as Meta

const PopoverStoryTemplate: Story<PopoverProps> = args => <Popover {...args} options={options} />

PopoverStoryTemplate.args = {
  onClick: action('onClick')
}

export const PopoverStory = PopoverStoryTemplate.bind({})
PopoverStory.args = {
  ...PopoverStoryTemplate.args
}
