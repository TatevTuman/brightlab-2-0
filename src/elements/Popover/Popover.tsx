import React, { memo } from 'react'
import { Icon } from '@elements'
import { PopoverOptionType } from '@types'
import './Popover.scss'

export interface PopoverProps {
  position?: string | undefined
  options?: PopoverOptionType[]
  open?: boolean
  onClick?(e: MouseEvent): void
}

const Popover: React.FC<PopoverProps> = props => {
  const { options, open, position, onClick } = props
  return (
    <div className={'popover'} data-open={open} position={position}>
      {options && (
        <div className={'popover-list'}>
          {options.length > 1 && <div className={'popover-list-line'} />}
          {options.map((option, index) => {
            const { label } = option
            return (
              <div key={index} className={'popover-list-option'} onClick={() => onClick}>
                <div>{label.name}</div>
                <div>{label.company}</div>
                <div>
                  <div>
                    <Icon name={'clock'} />
                  </div>
                  <div>{label.date}</div>
                </div>
              </div>
            )
          })}
        </div>
      )}
    </div>
  )
}
Popover.defaultProps = {
  open: true,
  options: [
    {
      label: {
        name: 'Michael Stevenson',
        company: 'Figma Corporate. Presentation and Sales...',
        date: '8:00 - 9:30'
      },
      value: 'First Option'
    },
    {
      label: {
        name: 'Michael Stevenson',
        company: 'Figma Corporate. Presentation and Sales...',
        date: '8:00 - 9:30'
      },
      value: 'Second Option'
    }
  ]
}

export default memo(Popover)
