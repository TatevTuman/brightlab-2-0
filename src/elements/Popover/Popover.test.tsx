import React from 'react'
import { act, cleanup, getByText, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Popover from './Popover'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Popover', () => {
  const props = {
    onClick: jest.fn(),
    options: [
      {
        label: {
          name: 'name-1',
          company: 'company-1',
          date: 'date-1'
        },
        value: 'test-1'
      },
      {
        label: {
          name: 'name-2',
          company: 'company-2',
          date: 'date-2'
        },
        value: 'test-2'
      }
    ]
  }
  it('renders correctly', async () => {
    const { container } = render(<Popover {...props} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
  })
  it('render option label correctly', () => {
    const { getByText } = render(<Popover {...props} />)

    props.options.forEach(({ label }) => {
      expect(getByText(label.name)).toBeInTheDocument()
      expect(getByText(label.company)).toBeInTheDocument()
      expect(getByText(label.date)).toBeInTheDocument()
    })
  })
  it('click', async () => {
    const { container } = render(<Popover {...props} />)
    const options = container.querySelectorAll('[key]')
    options.forEach(option => {
      act(() => {
        userEvent.click(option)
        expect(props.onClick).toHaveBeenCalledTimes(1)
      })
    })
  })
})
