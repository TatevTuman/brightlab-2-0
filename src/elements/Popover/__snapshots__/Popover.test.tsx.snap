// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders correctly 1`] = `
<div>
  <div
    class="popover"
    data-open="true"
  >
    <div
      class="popover-list"
    >
      <div
        class="popover-list-line"
      />
      <div
        class="popover-list-option"
      >
        <div>
          name-1
        </div>
        <div>
          company-1
        </div>
        <div>
          <div>
            <i
              class="icon"
            >
              <test-file-stub
                name="clock"
              />
            </i>
          </div>
          <div>
            date-1
          </div>
        </div>
      </div>
      <div
        class="popover-list-option"
      >
        <div>
          name-2
        </div>
        <div>
          company-2
        </div>
        <div>
          <div>
            <i
              class="icon"
            >
              <test-file-stub
                name="clock"
              />
            </i>
          </div>
          <div>
            date-2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
